core/src/main/java/hudson/Util.java

```diff
@@ -1071,7 +1072,12 @@ public static void touch(@NonNull File file) throws IOException {
 
     /**
      * Copies a single file by using Ant.
+     *
+     * @deprecated since TODO; use {@link Files#copy(Path, Path, CopyOption...)} directly
      */
+    @Deprecated
+    @Restricted(NoExternalUse.class)
```

Comment:
Please do not add `@Restricted` on top of `@Deprecated`. It is superfluous and can cause problems in cross-testing.

Reply:
the point of this pattern is to force consumers to update when they move their core version so it doesn't stay around forever. 

what's the problem in testing it causes? you could always run with ignoring restrictions for that

Reply:
> force consumers to update

There are a lot of things that are `@Deprecated` that consumers _should_ update as time permits as part of routine maintenance. If something is “worse than deprecated”—actively harmful and must be deleted soon in order to unblock something else—then seek out and fix callers. In this case there is no particular harm that I can see (unless and until we want to remove/detach Ant) so this annotation is just a nuisance.

The purpose of `@Restricted` is to mark a language element which must be `public` for some technical reason but which was never intended to be referred to from Java code, or only from Java code within the current project, etc.